scripts/ci/notification/matrix_notification.sh

```diff
@@ -0,0 +1,107 @@
+    # but it's broken. CircleCI associates runs on develop/breaking with random old PRs.
+    [[ "$BRANCH" == "develop" || "$BRANCH" == "breaking" ]] || { echo "Running on a PR or a feature branch - notification skipped."; exit 0; }
+
+    [[ -z "$1" ]] && fail "Event type not provided."
+    local event="$1"
+
+    # The release notification only makes sense on tagged commits. If the commit is untagged, just bail out.
+    [[ "$event" == "release" ]] && { [[ $TAG != "" ]] || { echo "Not a tagged commit - notification skipped."; exit 0; } }
+
+    workflow_name="$(circleci_workflow_name)"
+    job="$(circleci_job_name)"
+    formatted_message="$(format_message "$event")"
+
+    curl "https://${MATRIX_SERVER}/_matrix/client/v3/rooms/${MATRIX_NOTIFY_ROOM_ID}/send/m.room.message" \
+        --request POST \
+        --include \
+        --header "Content-Type: application/json" \
+        --header "Accept: application/json" \
+        --header "Authorization: Bearer ${MATRIX_ACCESS_TOKEN}" \
+        --data "$formatted_message"
+}
+
+circleci_workflow_name() {
+    # Workflow name is not exposed as an env variable. Has to be queried from the API.
+    # The name is not critical so if anything fails, use the raw workflow ID as a fallback.
+    workflow_info=$(curl --silent "https://circleci.com/api/v2/workflow/${CIRCLE_WORKFLOW_ID}") || true
+    workflow_name=$(echo "$workflow_info" | grep -E '"\s*name"\s*:\s*".*"' | cut -d \" -f 4 || echo "$CIRCLE_WORKFLOW_ID")
+    printf '%s' "$(sanitize_variable "$workflow_name")"
+}
+
+circleci_job_name() {
+    [[ $CIRCLE_NODE_TOTAL == 1 ]] && job="${CIRCLE_JOB}"
+    [[ $CIRCLE_NODE_TOTAL != 1 ]] && job="${CIRCLE_JOB} (run $((CIRCLE_NODE_INDEX + 1))/${CIRCLE_NODE_TOTAL})"
+    printf '%s' "$(sanitize_variable "$job")"
+}
+
+# Currently matrix only supports html formatted body,
+# see: https://spec.matrix.org/v1.6/client-server-api/#mtext
+#
+# Eventually, the matrix api will have support for better format options and `formatted_body` may not be necessary anymore:
+# https://github.com/matrix-org/matrix-spec-proposals/pull/1767
+format_message() {
+    [[ -z "$1" ]] && fail "Event type not provided."
+    local event="$1"
+
+    [[ "$event" == "failure" ]] && message="$(cat "${SCRIPT_DIR}/templates/build_fail.json")"
+    [[ "$event" == "success" ]] && message="$(cat "${SCRIPT_DIR}/templates/build_success.json")"
+    [[ "$event" == "release" ]] && message="$(cat "${SCRIPT_DIR}/templates/build_release.json")"
+
+    [[ -z "$message" ]] && fail "Message for event [$event] not defined."
+
+    # Escape \ and "
+    formatted_msg="$(printf '%s' "$message" | sed 's/\\/\\\\/g' | sed 's/"/\\"/g')"
+    # Evaluate environment variables
+    # NOTE: The variables must be sanitized before call eval to avoid execution
+    # of unexpected commands instead of valid input values.
+    # shellcheck disable=SC2005
+    echo "$(eval printf '%s' \""${formatted_msg}"\")"
```

Comment:
I don't think `eval` is such a good idea for template substitutions. I guess it gets the job done, but you have to worry about all that sanitization and stuff. And sanitization with regexes is never foolproof. I would not trust it without good tests (and creating them just for this would be a huge overkill).

Why not try something safer?
- There's an utility called [`envsubst`](https://www.gnu.org/software/gettext/manual/html_node/envsubst-Invocation.html) that's bundled with gettext. I bet we can assume we have it installed in CI most of the time.
- Alternatively, you can cobble up poor man's templating with a Python one-liner:
    ```bash
    python \
        -c 'import sys; print(open(sys.argv[1]).read().format(workflow_name="x"))' \ 
        build_success.json
    ```
    You'd just have to escape the braces in the JSON files (use `{{` and `}}` instead) and remove `$`.

    Not great but IMO still beats `eval` with hand-rolled sanitization.

Reply:
Right. Nice, I didn't know about `envsubst`. I replaced it and removed the `sanitization` stuff ;)